File: backend/app/services/dice/engine.py
"""Rolagem autoritativa no servidor.

Usamos ``secrets.SystemRandom`` (CSPRNG do sistema operacional): em mesa multiplayer ninguém
consegue prever ou forjar o resultado pelo cliente. Os testes injetam um ``random.Random`` com seed.
"""

import random
import secrets
from dataclasses import dataclass

from app.services.dice.notation import DiceExpression, DiceTerm

_system_rng = secrets.SystemRandom()
# Um dado que explode pode continuar rolando; o limite só evita laço infinito com um RNG viciado.
MAX_EXPLOSIONS = 20


@dataclass(frozen=True)
class DieResult:
    sides: int
    value: int
    kept: bool
    # Dados que explodem: cada rolagem (a soma é o value). Um só elemento = não explodiu.
    rolls: tuple[int, ...] = ()

    @property
    def first(self) -> int:
        """Resultado natural do primeiro lançamento (olhos de cobra, críticos)."""
        return self.rolls[0] if self.rolls else self.value

    def to_dict(self) -> dict:
        out = {"sides": self.sides, "value": self.value, "kept": self.kept}
        if len(self.rolls) > 1:
            out["rolls"] = list(self.rolls)
        return out


@dataclass(frozen=True)
class TermResult:
    term: DiceTerm
    dice: tuple[DieResult, ...]

    @property
    def subtotal(self) -> int:
        return self.term.sign * sum(d.value for d in self.dice if d.kept)


@dataclass(frozen=True)
class RollResult:
    expression: DiceExpression
    terms: tuple[TermResult, ...]

    @property
    def total(self) -> int:
        return sum(t.subtotal for t in self.terms) + self.expression.modifier

    def kept_positive_dice(self) -> list[DieResult]:
        """Dados que contam a favor (base da classificação de sucesso/falha)."""
        return [d for t in self.terms if t.term.sign > 0 for d in t.dice if d.kept]

    def to_dict(self) -> dict:
        return {
            "notation": self.expression.canonical(),
            "terms": [
                {
                    "notation": t.term.canonical(),
                    "sign": t.term.sign,
                    "dice": [d.to_dict() for d in t.dice],
                    "subtotal": t.subtotal,
                }
                for t in self.terms
            ],
            "modifier": self.expression.modifier,
            "total": self.total,
        }


def _keep_flags(values: list[int], term: DiceTerm) -> list[bool]:
    if term.keep is None:
        return [True] * len(values)
    # Empate: mantém o dado que saiu primeiro (ordem estável).
    order = sorted(range(len(values)), key=lambda i: (-values[i] if term.keep == "kh" else values[i], i))
    keep = set(order[: term.kept_count])
    return [i in keep for i in range(len(values))]


def roll_die(sides: int, explode: bool, rng: random.Random) -> tuple[int, ...]:
    rolls = [rng.randint(1, sides)]
    while explode and rolls[-1] == sides and len(rolls) <= MAX_EXPLOSIONS:
        rolls.append(rng.randint(1, sides))
    return tuple(rolls)


def roll(expression: DiceExpression, rng: random.Random | None = None) -> RollResult:
    rng = rng or _system_rng
    term_results = []
    for term in expression.terms:
        rolled = [roll_die(term.sides, term.explode, rng) for _ in range(term.count)]
        values = [sum(r) for r in rolled]
        flags = _keep_flags(values, term)
        dice = tuple(
            DieResult(sides=term.sides, value=v, kept=k, rolls=r if term.explode else ())
            for v, k, r in zip(values, flags, rolled, strict=True)
        )
        term_results.append(TermResult(term=term, dice=dice))
    return RollResult(expression=expression, terms=tuple(term_results))

File: mobile/src/lib/dice/engine.ts
import { canonical, DiceExpression, termToString } from './notation';
import type { RollPayload } from '../types';

/** Gerador 1..n. Offline usa Math.random; na mesa quem rola é o servidor (anti-trapaça). */
export type Rng = (sides: number) => number;

export const defaultRng: Rng = (sides) => 1 + Math.floor(Math.random() * sides);

function keepFlags(values: number[], keep: 'kh' | 'kl' | undefined, keepN: number | undefined): boolean[] {
  if (!keep || keepN === undefined) return values.map(() => true);
  // Empate: mantém o que saiu primeiro (mesma regra do backend).
  const order = values
    .map((v, i) => ({ v, i }))
    .sort((a, b) => (keep === 'kh' ? b.v - a.v : a.v - b.v) || a.i - b.i)
    .slice(0, keepN)
    .map((x) => x.i);
  const kept = new Set(order);
  return values.map((_, i) => kept.has(i));
}

export function rollLocal(expr: DiceExpression, rng: Rng = defaultRng): RollPayload {
  const terms = expr.terms.map((term) => {
    const values = Array.from({ length: term.count }, () => rng(term.sides));
    const flags = keepFlags(values, term.keep, term.keepN);
    const dice = values.map((value, i) => ({ sides: term.sides, value, kept: flags[i] ?? true }));
    const subtotal = term.sign * dice.filter((d) => d.kept).reduce((sum, d) => sum + d.value, 0);
    return { notation: termToString(term), sign: term.sign, dice, subtotal };
  });
  const total = terms.reduce((sum, t) => sum + t.subtotal, 0) + expr.modifier;
  return { notation: canonical(expr), terms, modifier: expr.modifier, total };
}

/** Pares [lados, valor] dos dados mantidos com sinal positivo (entrada da classificação). */
export function keptPositive(roll: RollPayload): [number, number][] {
  return roll.terms
    .filter((t) => t.sign > 0)
    .flatMap((t) => t.dice.filter((d) => d.kept).map((d) => [d.sides, d.value] as [number, number]));
}
